src/clj/fluree/db/query/exec/select.cljc

```diff
@@ -112,8 +122,15 @@
     [_ _ _ _ _ _ error-ch solution]
     (go (try* (agg-fn solution)
               (catch* e
-                (log/error e "Error applying aggregate selector")
-                (>! error-ch e))))))
+                      (log/error e "Error applying aggregate selector")
+                      (>! error-ch e)))))
+  (solution-value
+    [_ error-ch _]
+    (async/put!
+     error-ch
+     (ex-info "Aggregate selectors cannot be used in subqueries without 'as' syntax."
```

Comment:
Ideally we'd catch this in the parsing phase. We have enough information to be able to return an error before we even start reading the indexes. Malli should be able to detect this and throw before we execute any part of the query.

Reply:
Based on what I can figure out from Malli, the only way to do this is to create a new subquery definition that is mirrors query which is defined here: https://github.com/fluree/db/blob/f3dafc7d81c0741cb096304d32312747635e4a34/src/clj/fluree/db/query/fql/syntax.cljc#L35-L76  with a nearly identical function that duplicates everything except the two small tweaks to select.

This duplicationwoudl be a lot of duplication, but my larger concern is it is error-prone as any future features/syntax updates to query will have to be made to *both* definitions to keep them consistent.

Admittedly I'm no expert here, but I've spent a good deal of time here not only in the malli the docs, but also been in source code and don't see another path. If there is one I'd love to learn about it.

Reply:
I think we could refactor that function to make it have a minimal amount of duplication and allow future changes that apply to both query and subquery definitions only need to happen in one place because Malli schemas are just normal Clojure data. We could break out the common parts of the schema into it's own definitions, and then add to/transform them as necessary to create the specific schemas we need:

```suggestion
(def common-query-schema
  [:map {:closed true}
   [:from {:optional true} ::from]
   [:from-named {:optional true} ::from-named]
   [:where {:optional true} ::where]
   [:t {:optional true} ::t]
   [:context {:optional true} ::context]
   [:orderBy {:optional true} ::order-by]
   [:order-by {:optional true} ::order-by]
   [:groupBy {:optional true} ::group-by]
   [:group-by {:optional true} ::group-by]
   [:having {:optional true} ::function]
   [:values {:optional true} ::values]
   [:limit {:optional true} ::limit]
   [:offset {:optional true} ::offset]
   [:maxFuel {:optional true} ::max-fuel]
   [:max-fuel {:optional true} ::max-fuel]
   [:depth {:optional true} ::depth]
   [:opts {:optional true} ::opts]
   [:prettyPrint {:optional true} ::pretty-print]
   [:pretty-print {:optional true} ::pretty-print]])

(defn wrap-query-map-schema
  [schema]
  [:and
   [:map-of ::json-ld-keyword :any]
   [:fn {:error/fn
         (fn [_ _]
           (str "Query does not have exactly one select clause. "
                "One of 'select', 'selectOne', 'select-one', 'selectDistinct', or 'select-distinct' is required in queries. "
                "See documentation here for more details: "
                docs/error-codes-page "#query-missing-select"))}
    one-select-key-present?]
   schema])

(defn query-schema
  "Returns schema for queries, with any extra key/value pairs `extra-kvs`
  added to the query map.
  This allows eg http-api-gateway to amend the schema with required key/value pairs
  it wants to require, which are not required/supported here in the db library."
  [extra-kvs]
  (-> common-query-schema
      (into [[:select {:optional true} ::select]
             [:selectOne {:optional true} ::select]
             [:select-one {:optional true} ::select]
             [:selectDistinct {:optional true} ::select]
             [:select-distinct {:optional true} ::select]])
      (into extra-kvs)
      wrap-query-map-schema))

(defn subquery-schema
  [extra-kvs]
  (-> common-query-schema
      (into [[:select {:optional true} ::subquery-select]
             [:selectOne {:optional true} ::subquery-select]
             [:select-one {:optional true} ::subquery-select]
             [:selectDistinct {:optional true} ::subquery-select]
             [:select-distinct {:optional true} ::subquery-select]])
      (into extra-kvs)
      wrap-query-map-schema))
```
that 👆🏾 should work if `::subquery-select` is defined, and changes to any non-select part of the schema would automatically apply to both queries and subqueries. We could also change the `wrap-query-map-schema` function definition to take the error message and validator function as arguments, but it seems like the same function should work in both places so I opted not to do that.

Reply:
Resolved in d99b491bb69a12e2c0fd175d737005e4ca2419b9